api/envoy/api/v2/srds.proto

```diff
@@ -0,0 +1,214 @@
+service ScopedRoutesDiscoveryService {
+  rpc StreamScopedRoutes(stream DiscoveryRequest) returns (stream DiscoveryResponse) {
+  }
+
+  rpc IncrementalScopedRoutes(stream IncrementalDiscoveryRequest)
+      returns (stream IncrementalDiscoveryResponse) {
+  }
+
+  rpc FetchScopedRoutes(DiscoveryRequest) returns (DiscoveryResponse) {
+    option (google.api.http) = {
+      post: "/v2/discovery:scoped-routes"
+      body: "*"
+    };
+  }
+}
+
+// This configuration represents a set of routing 'scopes', each containing independent routing
+// configuration (see :ref:`routing architecture overview <arch_overview_http_routing>`).
+//
+// A ScopedRouteConfigurationsSet is associated with an
+// :ref:`HttpConnectionManager<envoy_api_msg_config.filter.network.http_connection_manager.v2.HttpConnectionManager>`
+// via the
+// :ref:`scoped_rds<envoy_api_field_config.filter.network.http_connection_manager.v2.HttpConnectionManager.scoped_rds>`
+// or
+// :ref:`scoped_routes_config<envoy_api_field_config.filter.network.http_connection_manager.v2.HttpConnectionManager.scoped_routes_config>`
+// fields. When scoped routing is enabled, the Envoy router will first attempt to determine the
+// routing scope assigned to a request by building a
+// :ref:`Scope.Key<envoy_api_msg_ScopedRouteConfigurationsSet.Scope.Key>` based on request header
+// attributes (such as the value of a header field) defined by the
+// :ref:`ScopeKeyBuilder<envoy_api_msg_ScopedRouteConfigurationsSet.ScopeKeyBuilder>`
+// this message.
+//
+// The key will be looked up in a table maintained by the router, containing the
+// :ref:`Scope.Key<envoy_api_msg_ScopedRouteConfigurationsSet.Scope.Key>` to
+// :ref:`RouteConfiguration<envoy_api_msg_RouteConfiguration>` mappings specified by the
+// :ref:`scopes<envoy_api_field_ScopedRouteConfigurationsSet.scopes>` field in this message. The
+// routing rules specified by the resulting RouteConfiguration will be used to determine the route
+// for the request.
+//
+// For example, with the following YAML configuration:
+//
+// .. code::
+//
+//   scope_key_builder:
+//     fragments:
+//       - header_element:
+//         name: X-Route-Selector
+//         element_separator: ,
+//         element:
+//           separator: =
+//           key: vip
+//   scopes:
+//     - route_configuration_name: route-config1
+//       key:
+//          fragments:
+//            - string_key: 172.10.10.20
+//      - route_configuration_name: route-config2
+//        key:
+//          fragments:
+//            - string_key: 172.20.20.30
+//
+// A request from a client such as:
+//
+// .. code::
+//
+//     GET / HTTP/1.1
+//     Host: foo.com
+//     X-Route-Selector: vip=172.10.10.20
+//
+// Would result in the `route-config1` RouteConfiguration being assigned to the request/stream,
+// which Envoy would fetch via RDS.
+//
+// [#comment:next free field: 4]
+message ScopedRouteConfigurationsSet {
```

Comment:
Upon revisiting the SRDS proto in the context of Incremental xDS, I have decided to make a few changes after discussing with @htuch.

To easily enable incremental updates to the routing scopes (represented by the `Scope` message) by leveraging the existing Incremental xDS API, I am going to move the `ScopeKeyBuilder` specification out of the `ScopedRouteConfigurationsSet` into the `HttpConnectionManager`, and instead of publishing a `repeated Scope scopes = ` in this message, each SRDS proto (which will most likely be renamed to `ScopedRouteConfiguration`) will contain a single `Scope`.

This change creates much better alignment with the API/protocol already defined for incremental updates, and also simplifies the SRDS API.

I am actively working on this now and will introduce the changes discussed as part of the next few commits to this PR.

Reply:
@AndresGuedez cool I will defer looking at the API again until this is done. Thank you for iterating on this.

/wait